### PR TITLE
added solution

### PR DESCRIPTION
**Pull request by VitaliiLeshchenko**

### CODE REVIEW

File: src/main/java/core/basesyntax/JavaDateTimeApi.java [hunk 1]
```diff
@@ -31,39 +44,46 @@ public String todayDate(DateTimePart datePart) {
      *                   - 3-й элемент массива - день (число);
      */
     public Optional<LocalDate> getDate(Integer[] dateParams) {
-        return Optional.empty();
+        if (dateParams.length == 0) {
+            return Optional.empty();
+        }
```
Comment (dmytro-hryhoruk) [inline src/main/java/core/basesyntax/JavaDateTimeApi.java:49]: do we need this check considering we are catching an exception?

File: src/main/java/core/basesyntax/JavaDateTimeApi.java [hunk 2]
```diff
@@ -31,39 +44,46 @@ public String todayDate(DateTimePart datePart) {
...
+        try {
+            return Optional.of(LocalDate.of(dateParams[0], dateParams[1], dateParams[2]));
+        } catch (DateTimeException e) {
+            return Optional.empty();
```
Comment (dmytro-hryhoruk) [inline src/main/java/core/basesyntax/JavaDateTimeApi.java:53]: put after catch block, we don't build logic on try catch blocks

File: src/main/java/core/basesyntax/JavaDateTimeApi.java [hunk 3]
```diff
@@ -74,7 +94,10 @@ public LocalDate addWeeks(LocalDate localDate, Integer numberOfWeeks) {
      * - "someDate is today" - если someDate - сегодня
      */
     public String beforeOrAfter(LocalDate someDate) {
-        return someDate + "is today";
+        LocalDate todayDate = LocalDate.now();
+        return someDate.toString() + (someDate.equals(todayDate) ? " is today"
+                        : someDate.isAfter(todayDate) ? " is after " + todayDate.toString()
+                        : " is before " + todayDate.toString());
```
Comment (dmytro-hryhoruk) [inline src/main/java/core/basesyntax/JavaDateTimeApi.java:100]: let's check isAfter() isBefore() if both false than consider is today without checking

File: src/main/java/core/basesyntax/JavaDateTimeApi.java [hunk 4]
```diff
@@ -96,23 +119,34 @@ public LocalDateTime getDateInSpecificTimeZone(String dateInString, String zone)
...
      * Дана строка в виде "yyyymmdd".
      * Необходимо вернуть Optional даты в LocalDate формате
      */
     public Optional<LocalDate> parseDate(String date) {
-        return Optional.empty();
+        try {
+            DateTimeFormatter dateTimeFormatter = DateTimeFormatter.ofPattern("yyyyMMdd");
+            return Optional.of(LocalDate.parse(date, dateTimeFormatter));
+        } catch (DateTimeException e) {
+            return Optional.empty();
```
Comment (dmytro-hryhoruk) [inline src/main/java/core/basesyntax/JavaDateTimeApi.java:134]: put after catch block, we don't build logic on try catch blocks

File: src/main/java/core/basesyntax/JavaDateTimeApi.java [hunk 5]
```diff
@@ -96,23 +119,34 @@ public LocalDateTime getDateInSpecificTimeZone(String dateInString, String zone)
...
      * Необходимо вернуть Optional даты в LocalDate формате
      */
     public Optional<LocalDate> customParseDate(String date) {
-        return Optional.empty();
+        try {
+            DateTimeFormatter dateTimeFormatter
+                    = DateTimeFormatter.ofPattern("d MMM yyyy", Locale.UK);
+            return Optional.of(LocalDate.parse(date, dateTimeFormatter));
+        } catch (DateTimeException e) {
+            return Optional.empty();
```
Comment (dmytro-hryhoruk) [inline src/main/java/core/basesyntax/JavaDateTimeApi.java:148]: same

**Review by dmytro-hryhoruk — COMMENTED** [review]
Make constants for Formatters

File: src/main/java/core/basesyntax/JavaDateTimeApi.java (hunk 2 above)
Comment (VitaliiLeshchenko) [reply src/main/java/core/basesyntax/JavaDateTimeApi.java]: dene

File: src/main/java/core/basesyntax/JavaDateTimeApi.java (hunk 5 above)
Comment (VitaliiLeshchenko) [reply src/main/java/core/basesyntax/JavaDateTimeApi.java]: done

File: src/main/java/core/basesyntax/JavaDateTimeApi.java (hunk 1 above)
Comment (VitaliiLeshchenko) [reply src/main/java/core/basesyntax/JavaDateTimeApi.java]: removed

File: src/main/java/core/basesyntax/JavaDateTimeApi.java [hunk 6]
```diff
@@ -96,14 +118,20 @@ public LocalDateTime getDateInSpecificTimeZone(String dateInString, String zone)
      * OffsetDateTime советуют использовать при записи даты в базу данных.
      */
     public OffsetDateTime offsetDateTime(LocalDateTime localTime) {
-        return OffsetDateTime.now();
+        return OffsetDateTime.of(localTime, ZoneOffset.ofHours(2));
```
Comment (dmytro-hryhoruk) [inline src/main/java/core/basesyntax/JavaDateTimeApi.java:121]: make constant for Offset

File: src/main/java/core/basesyntax/JavaDateTimeApi.java [hunk 7]
```diff
@@ -112,6 +140,13 @@ public OffsetDateTime offsetDateTime(LocalDateTime localTime) {
      * Необходимо вернуть Optional даты в LocalDate формате
      */
     public Optional<LocalDate> customParseDate(String date) {
+        try {
+            DateTimeFormatter dateTimeFormatter
+                    = DateTimeFormatter.ofPattern("d MMM yyyy", Locale.UK);
```
Comment (dmytro-hryhoruk) [inline src/main/java/core/basesyntax/JavaDateTimeApi.java:145]: make constants for all formatters)https://stackoverflow.com/questions/66066/what-is-the-best-way-to-implement-constants-in-java

File: src/main/java/core/basesyntax/JavaDateTimeApi.java [hunk 8]
```diff
@@ -123,6 +158,13 @@ public OffsetDateTime offsetDateTime(LocalDateTime localTime) {
...
     public String formatDate(LocalDateTime dateTime) {
-        return "Date can't be formatted!";
+        DateTimeFormatter dateTimeFormatter
+                = DateTimeFormatter.ofPattern("dd MMMM yyyy HH:mm", Locale.ENGLISH);
+        try {
+            return dateTime.format(dateTimeFormatter);
+        } catch (DateTimeException e) {
+            e.getMessage();
+        }
+        return null;
```
Comment (dmytro-hryhoruk) [inline src/main/java/core/basesyntax/JavaDateTimeApi.java:168]: we should return message "dateTime can't be formatted!" in this case

File: src/main/java/core/basesyntax/JavaDateTimeApi.java [hunk 9]
```diff
@@ -96,14 +124,20 @@ public LocalDateTime getDateInSpecificTimeZone(String dateInString, String zone)
...
+        return OffsetDateTime.of(localTime, ZoneOffset.ofHours(KYIV_TIME_ZONE));
     }
 
     /**
      * Дана строка в виде "yyyymmdd".
      * Необходимо вернуть Optional даты в LocalDate формате
      */
     public Optional<LocalDate> parseDate(String date) {
+        try {
+            DateTimeFormatter dateTimeFormatter = DateTimeFormatter.ofPattern(PATTERN_FOR_YYYYMMDD);
```
Comment (dmytro-hryhoruk) [inline src/main/java/core/basesyntax/JavaDateTimeApi.java:136]: probably we have ready-made formatter for this case?

File: src/main/java/core/basesyntax/JavaDateTimeApi.java [hunk 10]
```diff
@@ -112,6 +146,13 @@ public OffsetDateTime offsetDateTime(LocalDateTime localTime) {
      * Необходимо вернуть Optional даты в LocalDate формате
      */
     public Optional<LocalDate> customParseDate(String date) {
+        try {
+            DateTimeFormatter dateTimeFormatter
+                    = DateTimeFormatter.ofPattern(PATTERN_FOR_D_MMM_YYYY, Locale.UK);
```
Comment (dmytro-hryhoruk) [inline src/main/java/core/basesyntax/JavaDateTimeApi.java:151]: make all custom formatters as constants

File: src/main/java/core/basesyntax/JavaDateTimeApi.java [hunk 11]
```diff
@@ -1,12 +1,27 @@
...
+import java.time.ZoneOffset;
+import java.time.format.DateTimeFormatter;
+import java.util.Locale;
 import java.util.Optional;
 
 public class JavaDateTimeApi {
+
+    public static final String PATTERN_FOR_2D_MONTH_4D = "dd MMMM yyyy HH:mm";
+    public static final String PATTERN_FOR_D_MMM_YYYY = "d MMM yyyy";
+    public static final DateTimeFormatter FORMATTER1
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/JavaDateTimeApi.java:19]: FORMATTER1 і FORMATTER2 - не досить інформативні назви

File: src/main/java/core/basesyntax/JavaDateTimeApi.java [hunk 12]
```diff
@@ -96,14 +127,19 @@ public LocalDateTime getDateInSpecificTimeZone(String dateInString, String zone)
...
      * Дана строка в виде "yyyymmdd".
      * Необходимо вернуть Optional даты в LocalDate формате
      */
     public Optional<LocalDate> parseDate(String date) {
+        try {
+            return Optional.of(LocalDate.parse(date, DateTimeFormatter.BASIC_ISO_DATE));
+        } catch (DateTimeException e) {
+            e.getMessage();
+        }
         return Optional.empty();
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/JavaDateTimeApi.java:143]: також ми можемо перенести це в блок catch, аналогічно з іншими методами

File: src/main/java/core/basesyntax/JavaDateTimeApi.java [hunk 13]
```diff
@@ -16,9 +16,9 @@
 
     public static final String PATTERN_FOR_2D_MONTH_4D = "dd MMMM yyyy HH:mm";
     public static final String PATTERN_FOR_D_MMM_YYYY = "d MMM yyyy";
-    public static final DateTimeFormatter FORMATTER1
+    public static final DateTimeFormatter FORMATTER_2D_MONTH_4D
             = DateTimeFormatter.ofPattern(PATTERN_FOR_2D_MONTH_4D, Locale.ENGLISH);
-    public static final DateTimeFormatter FORMATTER2
+    public static final DateTimeFormatter FORMATTER_2D_MMM_YYYY
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/JavaDateTimeApi.java:21]:
```suggestion
    public static final DateTimeFormatter DATE_FORMATTER
```

File: src/main/java/core/basesyntax/JavaDateTimeApi.java [hunk 14]
```diff
@@ -16,9 +16,9 @@
 
     public static final String PATTERN_FOR_2D_MONTH_4D = "dd MMMM yyyy HH:mm";
     public static final String PATTERN_FOR_D_MMM_YYYY = "d MMM yyyy";
-    public static final DateTimeFormatter FORMATTER1
+    public static final DateTimeFormatter FORMATTER_2D_MONTH_4D
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/JavaDateTimeApi.java:19]:
```suggestion
    public static final DateTimeFormatter DATE_TIME_FORMATTER
```

**Review by Sofasmile — APPROVED** [review]

**Pull request closed without merging**